src/perfrecordssh.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "perfrecordssh.h"
+
+#include <QDir>
+#include <QFile>
+#include <QFileInfo>
+#include <QProcess>
+#include <QStandardPaths>
+
+#include <csignal>
+
+#include "hotspot-config.h"
+#include "ssh.h"
+
+PerfRecordSSH::PerfRecordSSH(QObject* parent)
+    : PerfRecord(parent)
+{
+}
+
+PerfRecordSSH::~PerfRecordSSH() = default;
+
+void PerfRecordSSH::record(const QStringList& perfOptions, const QString& outputPath, bool /*elevatePrivileges*/,
+                           const QString& exePath, const QStringList& exeOptions, const QString& workingDirectory)
+{
+    int exitCode = sshExitCode(m_deviceName, {QLatin1String("test"), QLatin1String("-e"), exePath});
+    if (exitCode) {
+        emit recordingFailed(tr("File '%1' does not exist.").arg(exePath));
+    }
+
+    exitCode = sshExitCode(m_deviceName, {QLatin1String("test"), QLatin1String("-f"), exePath});
+    if (exitCode) {
+        emit recordingFailed(tr("'%1' is not a file.").arg(exePath));
+    }
+
+    exitCode = sshExitCode(m_deviceName, {QLatin1String("test"), QLatin1String("-x"), exePath});
+    if (exitCode) {
+        emit recordingFailed(tr("File '%1' is not executable.").arg(exePath));
+    }
+
+    QStringList recordOptions = {exePath};
+    recordOptions += exeOptions;
+
+    startRecording(perfOptions, outputPath, recordOptions, workingDirectory);
+}
+
+void PerfRecordSSH::record(const QStringList& perfOptions, const QString& outputPath, bool /*elevatePrivileges*/,
+                           const QStringList& pids)
+{
+    if (pids.empty()) {
+        emit recordingFailed(tr("Process does not exist."));
+        return;
+    }
+
+    QStringList options = perfOptions;
+    options += {QStringLiteral("--pid"), pids.join(QLatin1Char(','))};
+    startRecording(options, outputPath, {}, {});
+}
+
+void PerfRecordSSH::recordSystem(const QStringList& perfOptions, const QString& outputPath)
+{
+    auto options = perfOptions;
+    options.append(QStringLiteral("--all-cpus"));
+    startRecording(options, outputPath, {}, {});
+}
+
+void PerfRecordSSH::stopRecording()
+{
+    if (m_recordProcess) {
+        m_userTerminated = true;
+        m_outputFile->close();
+        m_recordProcess->terminate();
+        m_recordProcess->waitForFinished();
+        m_recordProcess = nullptr;
+    }
+}
+
+void PerfRecordSSH::sendInput(const QByteArray& input)
+{
+    if (m_recordProcess)
+        m_recordProcess->write(input);
+}
+
+QString PerfRecordSSH::currentUsername()
```

Comment:
this is bogus, for ssh we provide the username externally (or not at all). and we don't allow runtime attaching via ssh anyways (or?). remove this for now

Reply:
can't do, since recordpage.cpp uses this to detect if off-cpu sampling is possible. But since off-cpu sampling is currently not possible via ssh I return an empty string.